File: python/tests/iroh_peer_child.py
"""Second OS process for the Iroh P2P test. Invoked by test_iroh.py."""

import json
import sys

from atomic_data import Store


def main() -> None:
    handshake_path, store_path, done_path = sys.argv[1:4]
    hs = json.loads(open(handshake_path, encoding="utf-8").read())
    store = Store.open(store_path)
    loaded = store.load_agent(hs["secret"])
    if loaded.drive_needs_sync:
        # Expected: this process has the agent but not the drive genesis yet.
        pass
    store.start_peer()
    report = store.sync_with(hs["node_id"], drive=hs["drive"])
    got = store.get(hs["subject"])
    open(done_path, "w", encoding="utf-8").write(
        json.dumps(
            {
                "ok": bool(got) and got["name"] == hs["name"],
                "imported": report.imported,
                "pushed": report.pushed,
                "name": got["name"] if got else None,
                "has_drive": store.has(hs["drive"]),
            }
        )
    )


if __name__ == "__main__":
    main()

File: ffi/kotlin/src/test/kotlin/dev/atomicdata/IrohPeerChild.kt
package dev.atomicdata

import java.io.File

/**
 * Second OS process for the Iroh P2P test. Invoked by [IrohTest].
 */
object IrohPeerChild {
    @JvmStatic
    fun main(args: Array<String>) {
        val handshakePath = args[0]
        val storePath = args[1]
        val donePath = args[2]
        val hs = parseObject(File(handshakePath).readText())
        val store = Store.open(storePath)
        store.loadAgent(hs.getValue("secret"))
        store.startPeer()
        val report = store.syncWith(hs.getValue("node_id"), hs["drive"])
        val got = store.get(hs.getValue("subject"))
        val ok = got != null && got.get("name") == hs["name"]
        File(donePath).writeText(
            buildString {
                append("{")
                append("\"ok\":").append(ok).append(',')
                append("\"imported\":").append(report.imported).append(',')
                append("\"pushed\":").append(report.pushed).append(',')
                append("\"in_sync\":").append(report.inSync).append(',')
                append("\"name\":").append(jsonString(got?.get("name"))).append(',')
                append("\"has_drive\":").append(store.has(hs.getValue("drive"))).append(',')
                append("\"child_peer\":").append(jsonString(store.peerId()))
                append("}")
            },
        )
    }

    private fun parseObject(raw: String): Map<String, String> {
        val body = raw.trim().removePrefix("{").removeSuffix("}")
        val out = mutableMapOf<String, String>()
        var i = 0
        while (i < body.length) {
            val keyStart = body.indexOf('"', i)
            if (keyStart < 0) break
            val keyEnd = body.indexOf('"', keyStart + 1)
            val colon = body.indexOf(':', keyEnd + 1)
            val valueStart = body.indexOf('"', colon + 1)
            val valueEnd = body.indexOf('"', valueStart + 1)
            out[body.substring(keyStart + 1, keyEnd)] = body.substring(valueStart + 1, valueEnd)
            i = valueEnd + 1
        }
        return out
    }

    private fun jsonString(value: String?): String =
        if (value == null) "null" else "\"${value.replace("\"", "\\\"")}\""
}
